src/main/app/controller/courts/AddressController.ts

```diff
@@ -0,0 +1,276 @@
+    const errors: string[] = [];
+
+    if (isPrimaryAddress || validateSecondaryAddress) {
+      if (!address.address_lines?.trim()) {
+        errors.push(this.addressRequiredError);
+      }
+      if(!address.town?.trim()) {
+        errors.push(this.townRequiredError);
+      }
+    }
+    return errors;
+  }
+
+  private validatePostcode(courtAddress: DisplayAddress, isPrimaryAddress: boolean): string[] {
+    const errors: string[] = [];
+    const postcodeRequired = isPrimaryAddress || (!isPrimaryAddress && this.addressFieldsNotEmpty(courtAddress));
+    const postcodeEmpty = !courtAddress?.postcode?.trim();
+
+    if (postcodeRequired && postcodeEmpty) {
+      errors.push(this.postcodeMissingError);
+    }
+
+    if (!postcodeEmpty && !this.postcodeIsValidFormat(courtAddress.postcode)) {
+      errors.push(this.invalidPostcodeError);
+    }
+
+    return errors;
+  }
+
+  private validateAddressTypes(primary: DisplayAddress, secondary: DisplayAddress, writeToUsTypeId: number): string[] {
+    return writeToUsTypeId && !!primary.type_id && !!secondary.type_id && // validate only if types selected in both addresses
+    (this.addressFieldsNotEmpty(secondary) || secondary.postcode?.trim()) && // validate only if secondary has some fields entered
+    (primary.type_id !== writeToUsTypeId && secondary.type_id !== writeToUsTypeId) // at least 1 write address should exist
+      ? [this.multipleVisitAddressError]
+      : [];
+  }
+
+  private addressFieldsNotEmpty(courtAddress: DisplayAddress): boolean {
+    return !!courtAddress.address_lines?.trim() || !!courtAddress.address_lines_cy?.trim() ||
+      !!courtAddress.town?.trim() || !!courtAddress.town_cy?.trim();
+  }
+
+  private postcodeIsValidFormat(postcode: string): boolean {
+    const postcodeRegex = new RegExp('([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|' +
+      '(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))' +
+      '\\s?[0-9][A-Za-z]{2})', 'g');
+
+    const match = postcode.match(postcodeRegex);
+    return match?.length === 1 && match[0] === postcode;
+  }
+
+  private getAddressTypesForSelect(addressTypes: AddressType[], isPrimaryAddress: boolean): SelectItem[] {
+    const allAddressTypes = addressTypes.map((at: AddressType) => ({ value: at.id, text: at.name, selected: false }));
+
+    return isPrimaryAddress
+      ? allAddressTypes
+      : allAddressTypes.filter(at => at.text.toLowerCase() !== this.visitOrContactUsAddressType.toLowerCase());
+  }
+
+  private checkErrorResponseForPostcodeErrors(error: AxiosError, addresses: DisplayCourtAddresses):
+    { primaryInvalid: boolean; secondaryInvalid: boolean; errors: string[] } {
+
+    let primaryPostcodeInvalid = false;
+    let secondaryPostcodeInvalid = false;
+    const errors: string[] = [];
+
+    // We expect an array of invalid postcodes in the body of the response
+    if (Array.isArray(error.response.data)) {
```

Comment:
We have an if statement here for if we have error.response.data, but no else. Should we have something in the event that there is no response data?

Reply:
This is called upon 400 Bad Request error responses from a PUT request to the API. We expect the invalid postcodes to be contained in an array in the response data. If that's not the case, we've encountered a 400 for some other reason and we let the general error handling take care of it.